Added iExec environment

KarowM opened the pull request:
Added the iExec environment along with a simple test class.

KarowM committed:
added iexec env

KarowM committed:
added iexec test file, removed hard coded path from iexec job description

jopasserat committed:
[iExec] enh: refactor examples to fit gridscale structure

jopasserat committed:
[iExec] add: iexec core and example to build infrastructure

jopasserat committed:
[iExec] fix: use LocalCluster interpreter in example

jopasserat committed:
[iExec] fix: commit scalariform formatting

KarowM committed:
Fixed spelling mistake, removed print statement

KarowM committed:
retrieveJobID processes stderr if stdout is empty
retrieveJobID fixed

KarowM committed:
retrieveJobID processes stderr if stdout is empty
retrieveJobID fixed

KarowM committed:
Simple working implementation for factorial

KarowM committed:
- Added path to iexec binaries in job description

jopasserat committed:
[iExec] fix: preserve existing PATH

KarowM committed:
- Removed walletPath from job description (not currently being used)

jopasserat committed:
[iExec] fix: use isEmpty to test stdout string

jopasserat commented on iexec/src/main/scala/gridscale/iexec/package.scala:
```diff
@@ -0,0 +1,83 @@
+      s"""export PATH="${jobDescription.IexecFilesPath}:$${PATH}" && cd ${jobDescription.workDirectory} && iexec result ${job.jobId} --dapp ${jobDescription.dappAddress} --save""",
+      impl.parseState)(server, job)
+
+  def clean[S](server: S, job: BatchJob)(implicit hn: HeadNode[S]): Unit = {
```
// TODO no cancel command yet indeed but i'd be surprised if it wasn't showing up at some point ;)

jopasserat commented on iexec/src/main/scala/gridscale/iexec/package.scala:
```diff
@@ -0,0 +1,83 @@
+         |#!/bin/bash
+         |export PATH="${description.IexecFilesPath}:$${PATH}"
+         |cd ${description.workDirectory}
+         |iexec account login
```
iExec maintains a session based on your wallet address => you shouldn't have to login everytime, you can call iexec account show first to see if you're logged in

jopasserat commented on iexec/src/main/scala/gridscale/iexec/package.scala:
```diff
@@ -0,0 +1,83 @@
+         |export PATH="${description.IexecFilesPath}:$${PATH}"
+         |cd ${description.workDirectory}
+         |iexec account login
+         |iexec account allow ${description.dappCost}
```
we will want a mode where you allow more than the dApp price to be able to submit multiple jobs without having to wait for a transaction to be mined everytime, so this should be a separate function
you can verify the account balance before submission (`iexec account show`) and compare it to the dapp price

later with a finer integration you'll be able to query the dApp's price from its smart contract directly

KarowM committed:
- extracted login and account allow from submission script
- iexec account needs to be populated before submission, allows for multiple job submission with a single account allow transaction
- verification ensuring account balance is at least as much as the dappCost

KarowM committed:
Merge remote-tracking branch 'origin/master'

KarowM committed:
- use isEmpty instead of comparison to empty string

jopasserat commented on iexec/src/main/scala/gridscale/iexec/package.scala:
```diff
@@ -0,0 +1,107 @@
+    def loginIexecAccount[S](server: S, description: IEXECJobDescription)(implicit hn: HeadNode[S], system: Effect[System]) = {
+      if (!new java.io.File(s"""/${description.workDirectory}/account.json""").exists) {
+        hn.execute(server, s"""cd ${description.workDirectory} && export PATH="${description.IexecFilesPath}:$${PATH}" && iexec account login""")
+      }
```
Would be good to find out whether the token generated with `iexec acount login` expires at some point? Maybe ask the team on Slack

jopasserat commented on iexec/src/main/scala/gridscale/iexec/package.scala:
```diff
@@ -0,0 +1,107 @@
+      val rlcBalance = ropstenBalanceLine.substring(ropstenBalanceLine.indexOf(":")+1, ropstenBalanceLine.indexOf(" nRLC")).trim.toInt
+
+      assert(rlcBalance >= description.dappCost, s"iExec account does not have enough tokens to execute this DApp, account needs at least ${description.dappCost} but has ${rlcBalance}")
+    }
```
Perfect! Now you just need another function to set the allowance. It would not be called automatically in submit but available to be called manually by the user.

jopasserat commented:
please build this

jopasserat commented:
please build this

KarowM committed:
- renamed attribute to follow naming convention

jopasserat committed:
Merge remote-tracking branch 'upstream/master'

jopasserat commented:
please build this

jopasserat commented:
please build this
